databuilder/dummy_data_generator.py

```diff
@@ -1,20 +1,239 @@
+            rows = get_rows(table)
+            for row in rows:
+                row["patient_id"] = self.patient_id
+                # Fill in any values that haven't already been set by a specialised
+                # generator
+                self.populate_row(table, row)
+            yield from (orm_class(**row) for row in rows)
+
+    def configure_next_patient(self):
+        self.patient_id += 1
+
+        date_of_birth = self.today - timedelta(days=self.rnd.randrange(0, 120 * 365))
+        age_days = self.rnd.randrange(105 * 365)
+        date_of_death = date_of_birth + timedelta(days=age_days)
+
+        self.date_of_birth = date_of_birth
+        self.date_of_death = date_of_death if date_of_death < self.today else None
+        self.events_start = self.date_of_birth
+        self.events_end = min(self.today, date_of_death)
+
+    def rows_for_patients(self, table):
+        row = {
+            "date_of_birth": self.date_of_birth.replace(day=1),
+            "date_of_death": self.date_of_death,
+        }
+        return [row]
+
+    def rows_for_practice_registrations(self, table):
+        # For now, just assume that every patient is permanently registered with a
+        # single practice from birth
+        row = {
+            "start_date": self.events_start,
+            "end_date": None,
+        }
+        return [row]
+
+    def empty_rows(self, table):
+        # Generate a small handful of events for event-level tables
+        max_rows = 1 if isinstance(table, SelectPatientTable) else 8
+        row_count = self.rnd.randrange(max_rows + 1)
+        return [{} for _ in range(row_count)]
+
+    def populate_row(self, table, row):
+        for column in self.inspector.get_columns(table):
+            # We only populate columns which haven't already been set
+            if column in row:
+                continue
+            values = self.inspector.get_values(table, column)
+            type_ = self.inspector.get_type(table, column)
+            row[column] = self.get_random_value(values, type_)
+
+    def get_random_value(self, values, type_):
+        if values:
+            return self.rnd.choice(values)
+        elif type_ is bool:
+            return self.rnd.choice((True, False, None))
+        elif type_ is int:
+            return self.rnd.randrange(100)
+        elif type_ is float:
+            return self.rnd.random() * 100
+        elif type_ is str:
+            # There's not much we can about non-categorical strings with no values we
+            # can find in the query, but we shouldn't have many of these
+            return ""
+        elif type_ is date:
+            # Use an exponential distribution to preferentially generate recent events
+            # (mean of one year ago)
+            days_ago = int(self.rnd.expovariate(1 / 365))
+            event_date = self.events_end - timedelta(days=days_ago)
+            # Clip to the available time range
+            return max(event_date, self.events_start)
```

Comment:
I'm really taken with this 🙂 As elsewhere, my main thought is "What if we wanted to use another distribution?" but you've already noted that there are some potential areas for improvement. If we did want to use another distribution, then would you try to accommodate it within this architecture? (e.g. `get_random_value` could be a visitor).